utils/litemap/src/serde.rs

```diff
@@ -76,58 +77,38 @@ where
     where
         S: SeqAccess<'de>,
     {
-        let mut map = LiteMap::with_capacity(access.size_hint().unwrap_or(0));
+        let mut kvs = Vec::with_capacity(access.size_hint().unwrap_or(0));
 
         // While there are entries remaining in the input, add them
         // into our map.
         while let Some((key, value)) = access.next_element()? {
-            // Try to append it at the end, hoping for a sorted map.
-            // If not sorted, insert as usual.
-            // This allows for arbitrary maps (e.g. from user JSON)
-            // to be deserialized into LiteMap
-            // without impacting performance in the case of deserializing
-            // a serialized map that came from another LiteMap
-            if let Some((key, value)) = map.try_append(key, value) {
-                // Note: this effectively selection sorts the map,
-                // which isn't efficient for large maps
-                map.insert(key, value);
-            }
+            kvs.push((key, value));
         }
 
-        Ok(map)
+        Ok(LiteMap::from_iter(kvs.into_iter()))
     }
 
     fn visit_map<M>(self, mut access: M) -> Result<Self::Value, M::Error>
     where
         M: MapAccess<'de>,
     {
-        let mut map = LiteMap::with_capacity(access.size_hint().unwrap_or(0));
+        let mut kvs = Vec::with_capacity(access.size_hint().unwrap_or(0));
 
         // While there are entries remaining in the input, add them
         // into our map.
         while let Some((key, value)) = access.next_entry()? {
-            // Try to append it at the end, hoping for a sorted map.
-            // If not sorted, insert as usual.
-            // This allows for arbitrary maps (e.g. from user JSON)
-            // to be deserialized into LiteMap
-            // without impacting performance in the case of deserializing
-            // a serialized map that came from another LiteMap
-            if let Some((key, value)) = map.try_append(key, value) {
-                // Note: this effectively selection sorts the map,
-                // which isn't efficient for large maps
-                map.insert(key, value);
-            }
+            kvs.push((key, value));
         }
 
-        Ok(map)
+        Ok(LiteMap::from_iter(kvs.into_iter()))
```

Comment:
Nit/Issue: This now allocates the vector twice.

Although we have gone away from using LiteMap in most performance-sensitive uses _in ICU4X_, LiteMap is still a good utility for the Rust ecosystem, and we shouldn't reduce its performance willy-nilly.

Reply:
Added back the serde implementation which used push.